pyiceberg/table/__init__.py

```diff
@@ -1178,10 +1178,15 @@ def update_table_metadata(
     """
     context = _TableMetadataUpdateContext()
     new_metadata = base_metadata
+    new_metadata = new_metadata.model_copy(update={"last_updated_ms": 0})
 
     for update in updates:
         new_metadata = _apply_table_update(update, new_metadata, context)
```

Comment:
I think it might be more intuitive for the `_apply_table_update` to update the `last_updated_ms` field as it updates the table metadata. 

For example, in this https://github.com/apache/iceberg-python/blob/3966263d935b92bcf98877ce9ba8ea26c08a283f/pyiceberg/table/__init__.py#L928-L934
there are checks to just return the table metadata without any updates

Reply:
I think we could rely on the `context` to determine if the metadata has been updated. In `_apply_table_metadata`, we only `add_update` to the context when we actually change the metadata.
Given that there are some cases that we update `last_updated_ms` within `_apply_table_metadata`, I think we could do something like:
```python
if context.has_changes() and base_metadata.last_updated_ms == new_metadata.last_updated_ms:
    new_metadata = new_metadata.model_copy(update={"last_updated_ms": datetime_to_millis(datetime.now().astimezone())})
```
We could add `has_changes()` that checks if the size of underlying `_updates` is greater than 0 to [TableMetadataUpdateConetxt](https://github.com/apache/iceberg-python/blob/d0bfb4a67bccd15c2d2e9baa481a5f33fe4c5220/pyiceberg/table/__init__.py#L886). 

I think one benefit of this approach is code simplicity - we do not need to update `last_updated_ms` in every `_apply_table_update`

Would love to hear thoughts on this!

Reply:
I agree with @HonahX. @kevinjqliu, what you are suggesting did cross my mind but I thought it would be better to avoid setting `last-updated-ms` at multiple places.

I was checking the how this is handled in Iceberg, here a metadata instance is created with `lastUpdatedMillis` set to null
https://github.com/apache/iceberg/blob/3c46178855ecec864e1ea9ff67392c0c96d49e0d/core/src/main/java/org/apache/iceberg/TableMetadata.java#L937

Then in the build method, `lastUpdatedMillis` is updated to current time if not already set. It also has several checks to determine if there are any updated in metadata
https://github.com/apache/iceberg/blob/3c46178855ecec864e1ea9ff67392c0c96d49e0d/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1431-L1438

Reply:
I like your approach @HonahX

> we only add_update to the context when we actually change the metadata.

I didn't know about this, cool

> We could add has_changes() that checks if the size of underlying _updates is greater than 0 to

+1

Is it necessary to check for `base_metadata.last_updated_ms == new_metadata.last_updated_ms`?

Reply:
I think we should check it, if individual `_apply_table_update` is setting last-updated-ms, we should honor it, otherwise set the current time as the value.

Personally, the approach in Iceberg (referenced above) seems cleaner to me.